aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -445,108 +525,76 @@ void cpu_flash_attention(
   accum_t* buf_data = buf.data_ptr<accum_t>();
   scalar_t* buf_reduced_data = is_reduced_type ? buf_reduced.data_ptr<scalar_t>() : nullptr;
 
-  // Buffer to store padding query
-  scalar_t* query_padding_ptr = nullptr;
-  std::unique_ptr<scalar_t[]> query_padding_data;
-  if (!headSize_even && need_pack) {
-    query_padding_data = std::make_unique<scalar_t[]>(num_thread * qSplitSize * eheadSize);
-    query_padding_ptr = query_padding_data.get();
-  }
-  // Buffer to store Key and Value after transforms
+  // Buffer to store padding query and packing key/value
   scalar_t* key_reorder_ptr = nullptr;
-  std::unique_ptr<scalar_t[]> key_reorder_data;
   scalar_t* value_reorder_ptr = nullptr;
-  std::unique_ptr<scalar_t[]> value_reorder_data;
-  int kv_padding_size = (kvSize - 1) / kvSplitSize * ekvSplitSize + ekvTail;
+  scalar_t* query_padding_ptr = nullptr;
+  int64_t kv_padding_size = (kvSize - 1) / kvSplitSize * ekvSplitSize + ekvTail;
+  at::Tensor key_t_reorder;
+  at::Tensor value_t_reorder;
+  at::Tensor qeury_t_padding;
   if (need_pack) {
-    key_reorder_data = std::make_unique<scalar_t[]>(batchSize * num_head * eheadSize * rkvSize);
-    key_reorder_ptr = key_reorder_data.get();
-    value_reorder_data = std::make_unique<scalar_t[]>(batchSize * num_head * kv_padding_size * rHeadSize);
-    value_reorder_ptr = value_reorder_data.get();
+    key_t_reorder = at::empty(
+      {batchSize, num_head, eheadSize, kvSize},
+      c10::CppTypeToScalarType<scalar_t>::value);
+    value_t_reorder = at::empty(
+      {batchSize, num_head, kv_padding_size, headSize},
+      c10::CppTypeToScalarType<scalar_t>::value);
+    key_reorder_ptr = key_t_reorder.data_ptr<scalar_t>();
+    value_reorder_ptr = value_t_reorder.data_ptr<scalar_t>();
```

Comment:
Probably has a bit more overhead this way, but sure. You could use `c10::DeviceArray` with the cpu allocator to get an allocation without the tensor overhead. Not a big deal though.

Reply:
Thank you very much for your suggestion. I have done some measurements before and found that the tensor creation overhead here has a negligible impact on the entire kernel.